Complete sprint 11

Evo-Eclipse opened the pull request:
Authentication & Profile Flow:
- Added ProgressHUD with UI blocking mechanism to provide visual feedback during the OAuth process
- Added request cancellation to prevent race conditions in OAuth2Service
- Created ProfileService and ProfileImageService for fetching user data and avatar
- Added thread safety with main thread assertions
- Added KVO NSKeyValueObservation API in WebViewViewController

Data Management:
- Added URLSession+objectTask extension for generic JSON decoding
- Migrated OAuth token storage from UserDefaults to secure KeychainWrapper
- Created storage classes with caching and expiration handling
- Improved error handling and logging across the application

UI Refactoring:
- Transitioned entirely from storyboard-based UI to programmatic implementation
- Created custom TabBarController class
- Updated SplashViewController to be implemented programmatically
- Refactored ProfileViewController to use Kingfisher for loading profile images
- Added loading indicators and fade transition animations

Project Structure:
- Restructured project architecture:
  - Created Application folder for app-level files
  - Moved controllers into a dedicated Controllers folder
  - Moved network-related files to a Network folder
  - Moved view components to a Views folder
- Added Constants.swift with application keys and URLs
- Updated window configuration in SceneDelegate
- Changed the path of Info.plist from Resources to Application

Evo-Eclipse committed:
Add ProgressHUD loading animations for auth flow

- Integrate ProgressHUD library for visual feedback during OAuth process
- Manage loading animation visibility throughout authentication flow
- Refactor AuthViewController with clear responsibility separation

Evo-Eclipse committed:
Add OAuth2Service race condition protection

- Add request cancellation to prevent race conditions
- Track last used authorization code to prevent duplicate requests
- Add thread safety with main thread assertions

Evo-Eclipse committed:
Add UI race condition protection

- Create UIBlockingProgressHUD helper class
- Disable user interaction on the window during operations
- Replace direct ProgressHUD usage with blocking wrapper

Evo-Eclipse committed:
Implement profile loading flow

- Added ProfileService and ProfileImageService for fetching user data and avatar
- Created storage classes with caching and expiration handling
- Refactored OAuth2Service to match standarts
- Updated splash screen flow to properly load profile before showing the main UI
- Implemented profile UI with real user data display
- Moved NetworkError to separate file and added error cases

Evo-Eclipse committed:
Update KVO API

- Replace deprecated KVO implementation with modern NSKeyValueObservation API in WebViewViewController

Evo-Eclipse committed:
Add generic object task and improve error handling

- Add URLSession+objectTask extension for generic JSON decoding
- Refactor services to use the new objectTask method
- Improve error handling and logging
- Update error alert UI in auth and profile flows

Evo-Eclipse committed:
Add Kingfisher for image loading

- Add Kingfisher dependency to the project
- Refactor profile image loading using Kingfisher
- Add loading indicator and fade transition animation

Evo-Eclipse committed:
Add KeychainWrapper for secure token storage

- Add SwiftKeychainWrapper dependency
- Migrate OAuth token storage from UserDefaults to Keychain

Evo-Eclipse committed:
Refactor TabBar to code

- Create custom TabBarController class
- Switch ProfileViewController from storyboard to code layout
- Configure tab bar items programmatically

Evo-Eclipse committed:
Update window configuration

- Set SceneDelegate class explicitly in AppDelegate
- Create window and set root view controller manually

Evo-Eclipse committed:
Update SplashViewController to code

- Remove SplashViewController from storyboard
- Implement UI elements programmatically in SplashViewController
- Set SplashViewController as initial window root controller
- Update navigation to AuthViewController

Evo-Eclipse committed:
Rebuild UI from code & Reorganize project structure

This commit represents a complete rewrite of the ImageFeed application's UI layer:

- Transitioned entirely from storyboard-based UI to programmatic implementation
- Restructured the overall project architecture by:
  - Creating an Application folder for app-level files
  - Moving controllers into a dedicated Controllers folder
  - Moving network-related files to a Network folder
  - Moving view components to a Views folder
- Added Constants.swift file with application keys and URLs
- Recreated TabBarViewController programmatically to replace storyboard version
- Redesigned navigation flow to use proper push/pop and modal presentation
- Eliminated storyboard dependencies, IB outlets and actions in favor of programmatic implementations
- Fixed issues with object lifecycles and proper initialization

Alex-rev-dev commented on ImageFeed/Controllers/WebViewViewController.swift:
```diff
@@ -90,9 +81,32 @@ final class WebViewViewController: UIViewController {
     private func loadAuthPage() {
         guard var urlComponents = URLComponents(string: WebViewConstants.unsplashAuthorizeURLString) else {
-            print("Web View Controller: Invalid service URL")
+            assertionFailure("Invalid service URL")
```
Использование подобных выражений не приветствуется внутри основного кода
приложения:
https://stevenpcurtis.medium.com/precondition-assert-fatal-error-or-guard-in-your-swift-code-5f9297658be0

Evo-Eclipse replied:
It's not recommended to release to production broken links with no way to clearly determine why components aren't working

Alex-rev-dev commented on ImageFeed/Controllers/ProfileViewController.swift:
```diff
@@ -41,14 +45,37 @@ final class ProfileViewController: UIViewController {
+                object: nil,
+                queue: .main
+            ) { [weak self] _ in
+                guard let self = self else { return }
```
Можно `guard let self else { return }`

Alex-rev-dev commented on ImageFeed/Services/ProfileImageService.swift:
```diff
@@ -0,0 +1,110 @@
+    
+    // MARK: - Structures
+    
+    struct UserResult: Decodable {
```
Можно пододвинуть в начало класса (перед `shared`), так код будет более структурированным :)

Alex-rev-dev commented on ImageFeed/Services/ProfileImageService.swift:
```diff
@@ -0,0 +1,110 @@
+    private func makeUserRequest(token: String, username: String) -> URLRequest? {
+        var components = URLComponents()
+        components.scheme = "https"
+        components.host = "api.unsplash.com"
```
Значение можно вынести в константу

Evo-Eclipse committed:
Minor Swift syntax improvements

- Simplified guard let statement in ProfileViewController
- Added .new option to progress observer in WebViewViewController
